### PR TITLE
feat(mock): first pass at mock implementation of Exchange

### PR DESCRIPTION
**Pull request by desa**
First pass at implementing an exchange. Figured I'd play around with the problem a little bit before deciding to make a bunch of changes to the interfaces and what not.

### CODE REVIEW

File: trading/trading.go [hunk 1]
```diff
@@ -67,10 +67,12 @@ func (p AbstractPortfolio) Value(prices Prices) float64 {
 
 // TODO: include enough data for other types of orders and shorts.
 type Order struct {
-	StopPrice  float64
-	LimitPrice float64
-	Volume     int
+	Type   string
```
Comment (d-sparks) [inline trading/trading.go]: Would it be easy to make this an enum/iota?

File: mock/exchange.go [hunk 2]
```diff
@@ -15,16 +23,63 @@ func NewExchange(seed float64) *Exchange {
...
+		switch order.Type {
+		case "limit":
+			if price >= order.Price && orderCost <= m.portfolio.CashUSD {
+				m.portfolio.Stocks[ticker] += order.Volume
+				m.portfolio.CashUSD -= orderCost
+			} else {
+				unfilledOrders = append(unfilledOrders, order)
+			}
+		case "stop":
+			if price <= order.Price {
```
Comment (d-sparks) [inline mock/exchange.go]: I'm ignorant about the types of orders we could have. But should we have both stop and limit for buy and sell orders?

Also, do I have this backwards? I would think if the price is higher than order.Price that is more favorable condition for selling. Like I want to protect against selling the stock for $0.

File: trading/trading.go (hunk 1 above)
Comment (desa) [reply trading/trading.go]: Yup. Super Easy

File: mock/exchange.go (hunk 2 above)
Comment (desa) [reply mock/exchange.go]: I'll look into the order types a bit more.

Looks like there's `Buy/Sell` variations of each of the order types.

https://www.thebalance.com/making-sense-of-day-trading-order-types-1031387

Should we implemented all of these or just do a simpler buy market / sell market for the first pass and add all the other types later? I'm open to either.

Comment (d-sparks) [reply mock/exchange.go]: We can start small.

I think I was thinking of a stop-limit order, where we would have a stop like this and a limit below the stop price where we would not sell below the limit (e.g., protect against selling at extremely low prices).

We can start with these, but we should at least add `Buy/Sell` to `trading.Order`?

**Comment by d-sparks** [comment]
LGTM!

File: mock/exchange.go (hunk 2 above)
Comment (desa) [reply mock/exchange.go]: I can add all the order types. Shouldn't be too difficult. I'll do that now.

Comment (desa) [reply mock/exchange.go]: Added a few more things here. It's becoming even more clear to me that we need a more sophisticated simulator.

**Comment by d-sparks** [comment]
Looking good!

**Comment by desa** [comment]
pr is definitely stale at this point

**Pull request closed without merging**